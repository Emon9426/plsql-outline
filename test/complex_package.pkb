/* 复杂包体测试 */
CREATE OR REPLACE PACKAGE BODY complex_package AS
  -- 实现计算奖金函数
  FUNCTION calculate_bonus(p_salary NUMBER, p_factor NUMBER) RETURN NUMBER IS
    v_bonus NUMBER;
    
    -- 嵌套函数：验证因子
    FUNCTION validate_factor(p_factor NUMBER) RETURN BOOLEAN IS
    BEGIN
      RETURN p_factor BETWEEN 0.1 AND 0.5;
    END validate_factor;
    
  BEGIN
    IF NOT validate_factor(p_factor) THEN
      RAISE invalid_value;
    END IF;
    
    v_bonus := p_salary * p_factor;
    RETURN v_bonus;
  END calculate_bonus;
  
  -- 实现处理员工过程
  PROCEDURE process_employees(p_emp_tab IN OUT t_employee_tab) IS
    v_index PLS_INTEGER;
    
    -- 嵌套过程：更新薪水
    PROCEDURE update_salary(emp IN OUT t_employee) IS
    BEGIN
      emp.salary := emp.salary * 1.1; -- 涨薪10%
    END update_salary;
    
  BEGIN
    v_index := p_emp_tab.FIRST;
    WHILE v_index IS NOT NULL LOOP
      -- 调用嵌套过程
      update_salary(p_emp_tab(v_index));
      v_index := p_emp_tab.NEXT(v_index);
    END LOOP;
  END process_employees;
  
  -- 实现格式化名称函数
  FUNCTION format_name(p_name VARCHAR2) RETURN VARCHAR2 IS
    v_formatted VARCHAR2(100);
    
    -- 深层嵌套函数：大写首字母
    FUNCTION capitalize(str VARCHAR2) RETURN VARCHAR2 IS
    BEGIN
      RETURN UPPER(SUBSTR(str, 1, 1)) || LOWER(SUBSTR(str, 2));
    END capitalize;
    
  BEGIN
    -- 处理姓名格式
    DECLARE
      v_names APEX_APPLICATION_GLOBAL.VC_ARR2;
      v_result VARCHAR2(100) := '';
    BEGIN
      v_names := APEX_STRING.SPLIT(p_name, ' ');
      FOR i IN 1..v_names.COUNT LOOP
        v_result := v_result || capitalize(v_names(i)) || ' ';
      END LOOP;
      RETURN TRIM(v_result);
    EXCEPTION
      WHEN OTHERS THEN
        RETURN p_name;
    END;
  END format_name;
  
  -- 注释中的关键字测试：BEGIN END FUNCTION PROCEDURE DECLARE
  /* 
   * 包体结束
   * 包含关键字：BEGIN 和 END
   */
END complex_package;
/
