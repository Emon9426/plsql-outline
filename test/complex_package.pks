/* 复杂包规范测试 */
CREATE OR REPLACE PACKAGE complex_package AS
  -- 常量声明
  MAX_VALUE CONSTANT NUMBER := 1000;
  
  -- 类型声明
  TYPE t_employee IS RECORD (
    id NUMBER,
    name VARCHAR2(100),
    salary NUMBER
  );
  
  TYPE t_employee_tab IS TABLE OF t_employee INDEX BY PLS_INTEGER;
  
  -- 异常声明
  invalid_value EXCEPTION;
  
  -- 函数声明
  FUNCTION calculate_bonus(p_salary NUMBER, p_factor NUMBER) RETURN NUMBER;
  
  -- 过程声明
  PROCEDURE process_employees(p_emp_tab IN OUT t_employee_tab);
  
  -- 嵌套函数声明
  FUNCTION format_name(p_name VARCHAR2) RETURN VARCHAR2;
  
  -- 注释中的关键字测试：BEGIN END FUNCTION PROCEDURE DECLARE
  /* 
   * 这是一个多行注释
   * 包含关键字：BEGIN 和 END
   * 以及 FUNCTION, PROCEDURE
   */
END complex_package;
/
