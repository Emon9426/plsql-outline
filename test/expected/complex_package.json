{
  "nodes": [
    {
      "type": "PACKAGE_BODY",
      "name": "complex_package",
      "declarationLine": 2,
      "beginLine": null,
      "exceptionLine": null,
      "endLine": null,
      "level": 1,
      "children": [
        {
          "type": "FUNCTION",
          "name": "calculate_bonus",
          "declarationLine": 4,
          "beginLine": 13,
          "exceptionLine": null,
          "endLine": 16,
          "level": 2,
          "children": [
            {
              "type": "FUNCTION",
              "name": "validate_factor",
              "declarationLine": 8,
              "beginLine": 9,
              "exceptionLine": null,
              "endLine": 11,
              "level": 3,
              "children": []
            }
          ]
        },
        {
          "type": "PROCEDURE",
          "name": "process_employees",
          "declarationLine": 23,
          "beginLine": 32,
          "exceptionLine": null,
          "endLine": 38,
          "level": 2,
          "children": [
            {
              "type": "PROCEDURE",
              "name": "update_salary",
              "declarationLine": 27,
              "beginLine": 28,
              "exceptionLine": null,
              "endLine": 30,
              "level": 3,
              "children": []
            }
          ]
        },
        {
          "type": "FUNCTION",
          "name": "format_name",
          "declarationLine": 42,
          "beginLine": 51,
          "exceptionLine": 62,
          "endLine": 65,
          "level": 2,
          "children": [
            {
              "type": "FUNCTION",
              "name": "capitalize",
              "declarationLine": 46,
              "beginLine": 47,
              "exceptionLine": null,
              "endLine": 49,
              "level": 3,
              "children": []
            }
          ]
        }
      ]
    }
  ],
  "metadata": {
    "sourceFile": "complex_package.pkb",
    "totalLines": 50,
    "maxNestingDepth": 3
  }
}